src/transform/src/predicate_pushdown.rs

```diff
@@ -84,357 +82,408 @@ impl PredicatePushdown {
         get_predicates: &mut HashMap<Id, HashSet<ScalarExpr>>,
     ) {
-        relation.visit_mut_pre(&mut |e| {
-            // TODO(#2592): we want to replace everything inside the braces
-            // with the single line below
-            // `self.action(e, &mut get_predicates);`
-            // This is so that you have a series of dependent views
-            // A->B->C, you want to push propagated filters
-            // from A all the way past B to C if possible.
-            // Before this replacement can be done, we need to figure out
-            // replanning joins after the new predicates are pushed down.
-            if let RelationExpr::Filter { input, predicates } = e {
+        // TODO(#2592): we want to replace everything inside the braces
+        // with the single line below
+        // `self.action(e, &mut get_predicates);`
+        // This is so that you have a series of dependent views
+        // A->B->C, you want to push propagated filters
+        // from A all the way past B to C if possible.
+        // Before this replacement can be done, we need to figure out
+        // replanning joins after the new predicates are pushed down.
+        match relation {
+            RelationExpr::Filter { input, predicates } => {
                 if let RelationExpr::Get { id, .. } = **input {
                     // We can report the predicates upward in `get_predicates`,
                     // but we are not yet able to delete them from the `Filter`.
                     get_predicates
                         .entry(id)
                         .or_insert_with(|| predicates.iter().cloned().collect())
                         .retain(|p| predicates.contains(p));
+                } else {
+                    input.visit1_mut(|e| self.dataflow_transform(e, get_predicates));
                 }
             }
-        });
+            RelationExpr::Get { id, .. } => {
+                // Purge all predicates associated with the id.
+                get_predicates
+                    .entry(*id)
+                    .or_insert_with(HashSet::new)
+                    .clear();
+            }
+            x => {
+                x.visit1_mut(|e| self.dataflow_transform(e, get_predicates));
+                // Prevent local predicate lists from escaping.
+                if let RelationExpr::Let { id, .. } = x {
+                    get_predicates.remove(&Id::Local(*id));
+                }
+            }
+        }
     }
 
-    /// Single node predicate pushdown
+    /// Predicate pushdown
+    ///
+    /// This method looks for opportunities to push predicates toward
+    /// sources of data. Primarily, this is the `Filter` expression,
+    /// and moving its predicates through the operators it contains.
+    ///
+    /// In addition, the method accumulates the intersection of predicates
+    /// applied to each `Get` expression, so that the predicate can
+    /// in principle be pushed through to a `Let` binding, or to the
```

Comment:
what is "in principle" about this? can't any predicates which the Gets agree on always be pushed down?

Reply:
Yeah, I'm not sure why those words are there. Lemme re-read and edit.